src/batched/utils.rs

```diff
@@ -0,0 +1,524 @@
+        seq: Sequence,
+    ) -> Option<(Sequence, i32)> {
+        self.send_command(Command::Received(batch, sender, seq))
+            .await
+            .map(|echoes| (seq, echoes))
+    }
+
+    /// Register echoes for many different payloads at once
+    pub async fn send_many<'a>(
+        &'a self,
+        batch: Digest,
+        from: PublicKey,
+        checks: impl IntoIterator<Item = Sequence> + 'a,
+    ) -> impl Stream<Item = (Sequence, i32)> + 'a {
+        checks
+            .into_iter()
+            .map(|seq| self.send(batch, from, seq))
+            .collect::<FuturesUnordered<_>>()
+            .filter_map(|x| async move { x })
+    }
+
+    /// Register a conflicting block from  a given remote peer
+    pub async fn conflicts(
+        &self,
+        batch: Digest,
+        peer: PublicKey,
+        sequence: Sequence,
+    ) -> Option<(Sequence, i32)> {
+        self.send_command(Command::Conflict(batch, peer, sequence))
+            .await
+            .map(|count| (sequence, count))
+    }
+
+    /// Register many conflicts for different sequences for a single peer
+    pub async fn many_conflicts<'a>(
+        &'a self,
+        batch: Digest,
+        peer: PublicKey,
+        sequences: impl IntoIterator<Item = Sequence> + 'a,
+    ) -> impl Stream<Item = (Sequence, i32)> + 'a {
+        sequences
+            .into_iter()
+            .map(|seq| self.conflicts(batch, peer, seq))
+            .collect::<FuturesUnordered<_>>()
+            .filter_map(|x| async move { x })
+    }
+
+    /// Internal helper to send commands to the agent
+    async fn send_command(&self, cmd: Command) -> Option<i32> {
+        let (tx, rx) = oneshot::channel();
+
+        if self.command.send((cmd, tx)).await.is_err() {
+            None
+        } else {
+            rx.await.ok()
+        }
+    }
+}
+
+impl Default for EchoHandle {
+    fn default() -> Self {
+        Self::new(128)
+    }
+}
+
+#[derive(Eq, Copy, Clone, Debug)]
+enum EchoStatus {
+    Conflict(PublicKey),
+    Okay(PublicKey),
+}
+
+impl PartialEq for EchoStatus {
+    fn eq(&self, other: &Self) -> bool {
+        match self {
+            Self::Conflict(skey) => match other {
+                Self::Conflict(okey) => skey == okey,
+                Self::Okay(okey) => skey == okey,
+            },
+            Self::Okay(skey) => match other {
+                Self::Conflict(okey) => skey == okey,
+                Self::Okay(okey) => skey == okey,
+            },
+        }
+    }
+}
+
+impl Ord for EchoStatus {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        match self {
+            Self::Conflict(key) => match other {
+                Self::Okay(okey) => key.cmp(okey),
+                Self::Conflict(okey) => key.cmp(okey),
+            },
+            Self::Okay(key) => match other {
+                Self::Okay(okey) => key.cmp(okey),
+                Self::Conflict(okey) => key.cmp(okey),
+            },
+        }
+    }
+}
```

Comment:
weird implementation of `Ord`, I made the following which fails
```rust
    #[test]
    fn check_cmp() {
        let key = keyset(1).next().unwrap();
        let mut set = BTreeSet::<EchoStatus>::new();
        assert!(set.insert(EchoStatus::Okay(key)));
        assert!(set.insert(EchoStatus::Conflict(key)));
    }
```
if that's intended, please a comment

Reply:
This is intented. The wanted behavior is that once a peer has acknowledged a sequence it can't retract its acknowledgement. Thus you can't replace an Okay(key) with a conflict Conflict(key).

Reply:
okay, I though it was somewhat intended; add a comment and we're good